### PR TITLE
Draft: New Feature: Define application tags from outside.

### PR DESCRIPTION
**Pull request by yguenduez**
This PR enables defining a tag class from outside - as it defaulted to an `universal` tag.

If you have `const-generics` feature enabled, you can define your TLV with e.g.:
``` Rust
const TAG_CLASS: u8 = TagClass::Application as u8;
let implicit_application_bool = Implicit::<bool, 2, TAG_CLASS>>::new(true);
```
Enums are not yet support as const parameters, that is why I used u8 as a placeholder for now.

The PR also changed the parsing of the derive macros. Now it is possible to define a tag class like this:
```
#[derive(asn1::Asn1Read, asn1::Asn1Write)]
struct SomeStruct<'a> {
    #[implicit(10, required, application)]
    a: &'a [u8],
}
```

If nothing is provided - it will default to the old behavior, i.e. `universal` tags.

This PR also only implements application tags. If `private` or `context-specific` tags are needed, I would open a new Issue/PR for this.

TL;DR:

- You can define tags from outside - either with derive macros or directly with the lib itself.
- only `application` tags are possible at the moment

Just let me know, if the PR direction is correct or not.

### CODE REVIEW

**Comment by alex** [comment]
Thanks for submitting this! Looks like the MSRV builds are unhappy, BUT: I'm planning on bumping the MSRV next week, so when that happens this PR should become a tiny bit simpler. I'm going to wait to review it until after that.

Thanks!

**Comment by alex** [comment]
Ok, the MSRV changes landed -- which means more conflicts, but hopefully makes this PR simpler :-)

**Comment by yguenduez** [comment]
> Ok, the MSRV changes landed -- which means more conflicts, but hopefully makes this PR simpler :-)

Rebased - and threw away all const-generic feature flags. But in order to make this draft work it'll need a MSRV of 1.59 - as default values for const-generic primitive types (i.e. u8) are allowed from this version on. Otherwise I need an other idea for implementation :)

File: src/parser.rs [hunk 1]
```diff
@@ -299,6 +299,22 @@ impl<'a> Parser<'a> {
         parse_single(tlv.data())
     }
 
+    /// This is an alias for `read_element::<Explicit<T, tag, 1>>` for use when
+    /// MSRV is < 1.51.
```
Comment (alex) [inline src/parser.rs:303]: Now that MSRV is >1.51, do we need these at all?

File: src/tag.rs [hunk 2]
```diff
@@ -3,13 +3,25 @@ use crate::parser::{ParseError, ParseErrorKind, ParseResult};
 use crate::writer::{WriteBuf, WriteResult};
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
-pub(crate) enum TagClass {
+pub enum TagClass {
     Universal = 0b00,
     Application = 0b01,
     ContextSpecific = 0b10,
     Private = 0b11,
```
Comment (alex) [inline src/tag.rs:10]: As written, this makes the `u8` values into a public API, I'm not sure we want that. Can we make this a "regular" enum rather than one with fixed values?

File: src/writer.rs [hunk 3]
```diff
@@ -119,6 +119,17 @@ impl Writer<'_> {
         self.write_tlv(tag, |dest| Writer::new(dest).write_element(val))
     }
 
+    /// This is an alias for `write_element::<Explicit<T, tag, 1>>` for use when
+    /// MSRV is <1.51.
```
Comment (alex) [inline src/writer.rs:123]: Same thought here.

File: src/lib.rs [hunk 4]
```diff
@@ -186,19 +186,61 @@ pub fn to_optional_default<'a, T: PartialEq>(v: &'a T, default: &'a T) -> Option
...
+/// considered a part of the supported API surface.
+#[doc(hidden)]
+pub const fn implicit_tag_application(tag: u32, inner_tag: Tag) -> Tag {
+    Tag::new(tag, TagClass::Application, inner_tag.is_constructed())
+}
+
+/// This API is public so that it may be used from macros, but should not be
+/// considered a part of the supported API surface.
+#[doc(hidden)]
+pub const fn implicit_tag_context_specific(tag: u32, inner_tag: Tag) -> Tag {
```
[the remark this answers is not in the document]
Comment (yguenduez) [reply src/lib.rs]: No, true - I also removed the u8 from the generic parameters and changed the signature like so:
```rust
implicit_tag_class(tag: u8, tag_class: TagClass, ..)
```
adding the TagClass as parameter instead of a const-generic u8

File: src/parser.rs (hunk 1 above)
Comment (yguenduez) [reply src/parser.rs]: No we actually do not. These are used int the derive crate though. But I also exchanged them with a more generic part, that take a TagClass as parameter.

File: src/tag.rs (hunk 2 above)
Comment (yguenduez) [reply src/tag.rs]: I used them, as the Implicit/Explicit do only support primitve const generic types. I now have a const fn method, that returns an u8 on a regular enum. 

I do not have a better Idea right now - if you have, just let me know.

**Comment by yguenduez** [comment]
Thanks for the review! I changed the things you have linted. But I don't know what to do about the Implicit and Explicit generic ones. As only primitives are supported right now for const generics, my only solution I come up is something like:

```rust
const APPLICATION: TagClass = TagClass::Application;
Implicit::<Some(false), 42, { APPLICATION.as_u8() }>::new(...);
```

with `as_u8()` being a const method on the regular enum. But that does not really solve the `u8` issue. Guess that will only work, if enums are supported with const generics, which would then look like:

```rust
Implicit::<Some(false), 42,  TagClass::Application>::new(...);
```

File: asn1_derive/src/lib.rs [hunk 5]
```diff
@@ -228,19 +228,46 @@ enum OpType {
...
+            let required = &ident == "required";
+
+            if input.lookahead1().peek(syn::Token![,]) {
+                input.parse::<syn::Token![,]>()?;
+                tagclass =
+                    parse_tag_class_from_ident(input.parse::<syn::Ident>()?.to_string().as_str())
+                        .unwrap();
+            };
+
+            (tagclass, required)
```
Comment (alex) [inline asn1_derive/src/lib.rs:261]: I think this will silently allow nonsense, right?

File: src/types.rs [hunk 6]
```diff
@@ -1453,12 +1453,17 @@ impl<'a, T: Asn1Writable, V: Borrow<[T]>> SimpleAsn1Writable for SetOfWriter<'a,
 /// `Implicit` is a type which wraps another ASN.1 type, indicating that the tag is an ASN.1
 /// `IMPLICIT`. This will generally be used with `Option` or `Choice`.
 #[derive(PartialEq, Eq, Debug)]
-pub struct Implicit<'a, T, const TAG: u32> {
+pub struct Implicit<
+    'a,
+    T,
+    const TAG: u32,
+    const TAG_CLASS: u8 = { TagClass::ContextSpecific as u8 },
```
Comment (alex) [inline src/types.rs:1460]: I'm confused, how does this work? Surely it needs to be `.as_u8()`?

**Review by alex — COMMENTED** [review]
Didn't review in super detail. There's more merge conflicts :-( Sorry

**Comment by alex** [comment]
I should say: I'm happy with how this is coming together. Thanks for sticking with this!

**Comment by yguenduez** [comment]
Sorry, I do not have time in the near future to convert this draft into a pr. But I am thankful anyway, that you put your precious time into this review.

**Comment by alex** [comment]
Thank you for working on it, hopefully eventually I'll get it over the finish line.
